Issue/3039 modules v2 document dev to prod workflow (#3058)
inmanta/inmanta-core: docs/model_developers/modules.rst

 * Or a repository location such as  git+https://github.com/project/python-foo The correct syntax
   to use is then: eggname@git+https://../repository#branch with branch being optional.

Working on modules
==================
Since modules often have dependencies on other modules, it is common to develop against multiple
modules (or a project and one or more modules) simultaneously. One might for example need to
extend a dependency module to add support for some new feature. Because this use case is so common,
this section will describe how to work on multiple modules simultaneously so that any changes are
visible the compiler. This procedure is of course applicable for working on a single module as well.

> Setting up the dev environment

Reviewer: My general comment on this section is that we should make this have less `branches`.

This is external documentation, so I think we need an explanation that allows one to copy paste some  command and end up with a dev environment, without having to know exactly what is what. 

We can then in second instance elaborate on alternatives and nuances. 

I would assume, in first instance, that we are working in the context of a project, in a venv (as is in the developer getting started docs iirc), with a mix of packaged and unpackaged modules. 

In second instance we can add nuances such as e.g. having a single libs folder for multiple projects (by updating the search path in the project file)

Author: All right, I'll see what I can do. To what extent should I account for v1 modules in this initial focus?

Reviewer: The process should preferably be identical for both.

------------------------------
v1 modules
^^^^^^^^^^

> A v1 module is a module that is not packaged and published to a pip index but is solely distributed

Author: I prefer "Python package repository", even though it's long. I even think I used the exact same term in some places but I missed this one.

> A v1 module is a module that is not packaged and published to a pip index but is solely distributed

Reviewer: I think this is not the description of a v1 module, but of a 'from source' module? 

(not the best name, but what I mean is that this is about packaged/unpackaged, not about source layout)

> A v1 module is a module that is not packaged and published to a pip index but is solely distributed

Author: Could you elaborate? I don't think I agree. The next section talks about v2 modules in both packaged and source forms. This section is actually meant to be about v1.

> A v1 module is a module that is not packaged and published to a pip index but is solely distributed

Reviewer: I think this is mostly supersede by my general comment

via git. V1 modules are installed on the fly by the compiler and can be found in the project's

> `libs` directory. Any changes you make to the module source in the `libs` dir will be reflected in

Author: Good suggestion, will fix.

the next compile. If you need to be able to make changes to multiple modules simultaneously, the
simplest way to set up your development environment is to create a project that imports the desired
modules, then run a compile against it. Alternatively, you can manually clone the module repos.

v2 modules
^^^^^^^^^^
A v2 module in development form has mostly the same structure as a v1 module. The main difference is
that it is meant to be published as a Python package. A project or another v2 module then lists all

> required v2 modules as dependencies in its `pyproject.toml`. The compiler does not install v2

Reviewer: Why not? 

Doesn't the install behavior depend on the project setup: if you give it module sources of type 'git' or if you give it module source of type 'pypi'

(differentiating the install behavior of V1 and V2 is a bit hard, as you don't know up front what it is going to be)

Author: I guess this is something we need to discuss further then (which means this writing the documentation first approach is working). My takeaway from the design discussions was that we wanted to mimic the Python behavior in this regard: a module does not install its own dependencies, instead they must be installed when setting up the environment.
- it's simple in concept, which makes it easy to understand and to implement. No need to think about when and where the compiler installs it and whether it can be persisted in the compiler venv. Single install during env setup and that's it.
- the dependencies are essentially Python dependencies, which is a point in favor of this approach: we stay consistent with the Python conventions.
- it requires an extra step of actually installing the module (which pulls in all dependencies), which is a potential negative for this approach.

@arnaudsjs @bartv this is a design discussion which is probably relevant to all of us.

Reviewer: I'm not so much interested in installing a module, but in installing a project (we currently have no 'installing a module' per-se)

Author: Well, the same goes for a project in my interpretation.

Author: I agree that when you install a Python package it installs its dependencies. That would be the case here as well. What a Python project doesn't do is install its dependencies implicitly when you run/import the module. You need to actively install a package to set up the required environment. If you later remove one of its dependencies from the venv, it doesn't get reinstalled automatically except if you run `pip install .` again.

Author: I feel like most other discussions on this pull request depend on this one (some more closely than others). I think this is a pretty fundamental building block of the design. Perhaps we could discuss how we want this behavior in some more detail in tommorow's meeting? I feel like there are two major options:
- The compiler can implicitly install v2 modules via pip if they are not installed in the parent env. This behavior is close to the current behavior which makes it easy to understand. More difficult perhaps is the management of the v2 modules: does the compiler install it in its own venv? What then if the developer installs one of the modules in editable form in the parent venv at some later stage?
- The compiler expects all dependencies to be preinstalled (running `inmanta project install` or `inmanta modules install` could be sufficient). This puts the responsibility for module installation with the developer, allowing them to be more flexible and in control of which dependencies (packaged vs editable, exact version, ...) are installed. Additionally, I feel like this is the more natural approach if we label modules as Python packages, though I must admit the differences with standard Python packages are sufficient that this last argument might not carry that much weight.

Do you agree it would be a good idea to discuss this in a meeting in the hopes of reaching a consensus there before proceeding with the design details that follow from it?

Author: (I've context switched to the machine learning project mostly so this is not blocking me in the meantime. I don't think it is blocking the modules v2 in general either for now.)

Reviewer 2: Indeed. There are some loose ends here. I am in favor of a meeting.

modules on the fly. In line with how Python depdencies work in general they are expected to be
installed in advance. As a result they will not be placed in the project's `libs` directory.

> To set up your development environment, first clone all modules you wish to develop against and

Reviewer: Shouldn't we simply provide a command like `inmanta modules install` that installs everything?
And e.g. `inmanta modules dev-install` that also installs all dev dependencies? 

Or is this in a context outside of a specific project? 

also, it this installation in a normal venv or the compiler venv?

Author: - this is either in the context of a project or the context of a single module. I think we will indeed need something like `inmanta module install` and `inmanta project install` if we need to wrap `pip install -e`. For an editable install of a dependency I think we'll always need to add manual steps though.
- I was thinking of the development env outside of the compiler venv. I'd see these dependencies as something managed by the model developer (he can choose to do editable installs for example), so I don't think it would be right to put it in the compiler venv.

Reviewer: this command exists and is called `inmanta module install` (but that installs in the compiler venv and doesn't install dev dependencies).

install them in editable mode with `poetry install`. This will fetch all its module v2 dependencies

> from the Python package index and install them in the active Python environment. Any modules you

Reviewer 2: Shouldn't we mention that the user should update the `pyproject.toml` file in the following way when a editable install is required on a project:

```
[tool.poetry.dependencies]
my-package = {path = "../my/path", develop = true}
```

So point to the source directory instead of a python package and set the develop flag to true.

Author: Since we're moving away from poetry I assume I can disregard this comment?

Reviewer 2: yes

previously installed in editable mode should remain as is, provided they meet the version
constraints. Then optionally clone the project repo and install its other dependencies with
`poetry install`. You can double-check the desired modules are installed in editable mode by
checking the output of `pip list --editable`. If you want to add another module to the set under
development, you can always run `poetry install` on it at a later stage, overwriting the published

> package that was installed previously.

Reviewer 2: Executing `poetry install` calls on project and modules on the same venv in quite tricky, since `poetry install` takes into account the `poetry.lock` file. Example:

* Run `poetry install` on Inmanta project -> Create a poetry.lock file
* Run `poetry install` on module -> Creates a poetry.lock file in the module

Now you have two separate lock file. The next time you execute `poetry install` on the project again, you might alter the install done on the module, because its lock file mentions another version for example. I don't think `poetry` is designed to run one venv against multiple distinct projects.

Author: Since we're moving away from poetry I assume I can disregard this comment? If I replace all occurrences of `poetry install` with `pip install -e` this should be resolved I think.

Reviewer 2: Yes. I think so.

 * Or a repository location such as  git+https://github.com/project/python-foo The correct syntax
   to use is then: eggname@git+https://../repository#branch with branch being optional.

Working on modules
==================
Since modules often have dependencies on other modules, it is common to develop against multiple
modules (or a project and one or more modules) simultaneously. One might for example need to
extend a dependent module to add support for some new feature. Because this use case is so common,
this section will describe how to work on multiple modules simultaneously so that any changes are
visible to the compiler. This procedure is of course applicable for working on a single module as well.

Setting up the dev environment
------------------------------
To set up the development environment for a project, activate your development Python environment and
install the project with ``inmanta project install``. To set up the environment for a single module,

> run ``inmanta module install`` instead.

Author: Though we can just extend the behavior, no? When run in a module directory it just installs the module?

> run ``inmanta module install`` instead.

Reviewer: ok!

> The following subsections explain any additional steps you need to take if you want to make changes
> to one of the dependency modules as well.

Reviewer 2: ```suggestion
The following subsections explain any additional steps you need to take if you want to make changes
to one of the dependent modules as well.
```

v1 modules
^^^^^^^^^^
Any modules you find in the project's ``modulepath`` after starting from a clean project and setting
up the development environment are v1 modules. You can make changes to these modules and they will
be reflected in the next compile. No additional steps are required.

v2 modules
^^^^^^^^^^

> All other modules are v2 and have been installed by ``inmanta project install`` into the active Python

Author: Would we allow this? I thought v2 modules would be strictly Python packages, either from a package repo or from an editable install.

> All other modules are v2 and have been installed by ``inmanta project install`` into the active Python

Reviewer: yes, but an editable install may also result from a download in git form? 

(also: how can the compiler know if it is a V1 or V2 module prior to downloading it?)

> All other modules are v2 and have been installed by ``inmanta project install`` into the active Python

Author: 1. I think this is another thing we never discussed explicitly. According to my view an editable install would always be performed explicitly by the model developer. I thought we wanted to fully move away from the git distribution method for v2, and only support it for backwards compatibility with v1 modules.
2. I don't think it can. I think the repo needs to be cloned to determine this.

environment. If you want to be able to make changes to one of these modules, the easiest way is to

> check out the module repo separately and run ``inmanta module install`` on it, overwriting the published
> package that was installed previously This will install the module in editable form: any changes you make
> to the checked out files will be picked up by the compiler. You can also do this prior to installing the

Reviewer 2: ```suggestion
check out the module repo separately and run ``inmanta module install`` on it, overwriting the published
package that was installed previously. This will install the module in editable form: any changes you make
to the checked out files will be picked up by the compiler. You can also do this prior to installing the
```

------------------------------
After setting up, you should be left with a dev environment where all required v2 modules have been
installed (either in editable or in packaged form). If you're working on a project, all required v1
modules should be checked out in the ``modulepath`` directory.

When you run a compile from the active Python environment context, the compiler will find both the
v1 and v2 modules and use them for both their model and their plugins.

Similarly, when you run a module's unit tests, the installed v2 modules will automatically be used
by the compiler. As for v1 modules, by default, the ``pytest-inmanta`` extension makes sure the
compile itself runs against an isolated project, downloading any v1 module dependencies at
compile-time. If you want to compile against local versions of v1 modules, have a look at the
``--use-module-in-place`` option in the ``pytest-inmanta`` documentation.

Distributing modules
====================
This section is about v2 modules. V1 modules only require a version tag to be recognized as a
released version. While a version tag is still good practice for v2 modules, it isn't sufficient
to consider it released.

> You can package a v2 module with ``inmanta module build`` which will create an sdist and a bdist

Reviewer 2: * The design document mentions that we will only create wheels. Maybe we should mention that here explicitly.
* I don't think it's a good idea to build binary packages. Wouldn't it be sufficient to build universal wheels and leave the compilation to the server and agent? @wouterdb What is your opinion on this?

Reviewer: - I have nothing against sdist  (I don't know the argument around this)
- I don't see how we would get anything binary-ish, so I would assume our wheels work on any python interpreter?

Author: I'm not too familiar with this. I mainly wrote this because this is what we usually generate for our Python packages as far as I know. I'll update it to state we build a wheel.

of the Python package. You can then publish this to the Python package repository of your choice,
for example the public PyPi repository. For an inmanta project, follow the same procedure but

> substitute ``module`` with ``project``.

Reviewer: are we also going to support packaging projects already? (I guess it is not that much extra work)

Author: I understood this to be part of the MVP, but that might be a misunderstanding.

Reviewer: it somewhat contradicts the 'M' in MVP, but we can support it.

The orchestrator server generally (see

> :ref:`Advanced concepts<modules-distribution-advanced-concepts`) installs both project and modules

Reviewer 2: This reference doesn't seems to works.

Reviewer 2: ```suggestion
:ref:`Advanced concepts<modules-distribution-advanced-concepts>`) installs both project and modules
```

from the configured Python package repository, respecting the environment's version constraints on

> the project package, the project's constraints on its modules and all inter-module constraints. The

Reviewer 2: What is the difference between these two:
* The environment's version constraints on the project package
* The project's constraints on its modules

Author: I'll get rid of this. I wrote this thinking the project would be a Python project as well, but your pull request showed that it wouldn't. I forgot to change it afterwards.

> server is then responsible for supplying the agents with the appropriate ``inmanta_plugins``
> modules.

Reviewer 2: ```suggestion
server is then responsible for supplying the agents with the appropriate ``inmanta_plugins``
packages.
```

.. _modules-distribution-advanced-concepts

Advanced concepts
-----------------

Freezing a project
^^^^^^^^^^^^^^^^^^
Prior to releasing a new stable version of an inmanta project, you might wish to freeze its module
dependencies. This will ensure that the orchestrator server will always work with the exact
versions specified. You can achieve this with

> ``inmanta project freeze --recursive --operator "=="``. This command will freeze all module
> depdencies to their exact version as they currently exist in the Python environment. The recursive
> option makes sure all module depdencies are frozen, not just the direct dependencies. In other
> words, if the project depdends on module ``a`` which in turn depends on module ``b``, both modules
> will be pinned to their current version in ``pyproject.toml``.

Reviewer 2: ```suggestion
``inmanta project freeze --recursive --operator "=="``. This command will freeze all module
dependencies to their exact version as they currently exist in the Python environment. The recursive
option makes sure all module dependencies are frozen, not just the direct dependencies. In other
words, if the project depends on module ``a`` which in turn depends on module ``b``, both modules
will be pinned to their current version in ``pyproject.toml``.
```

Author: Wow, this really shows an issue with my typing...

> will be pinned to their current version in ``pyproject.toml``.

Reviewer: ```suggestion
will be pinned to their current version in ``setup.cfg``.
```

Manual export
^^^^^^^^^^^^^

> The `inmanta export` command exports a project and all its modules to the orchestrator server.

Reviewer 2: ```suggestion
The ``inmanta export`` command exports a project and all its modules to the orchestrator server.
```

When this method is used, the orchestrator does not install any modules from the Python package

> repository but instead contains all code (both model and plugins) as present in the local Python

Reviewer: inmanta export sends no model at all (only resources and python code) ?

Author: I think I misunderstood something then. Last week we discussed two modes of the server: pull from version control and manual export of a model. With the second mode I was referring to this, which is apparently not correct. What then is this second mode?
In any case, I'll update this to say it exports the Python code only.
